vsphere/vsphere.go

```diff
@@ -398,6 +532,476 @@ func (ops *vsphereOps) Tags(volumeID string) (map[string]string, error) {
+				Name: vm.Name(),
+				ID:   vmUUID,
+				Zone: vmZone,
+			},
+		})
+	}
+
+	return resp, nil
+}
+
+func (ops *vsphereOps) getUUIDForVM(ctx context.Context, vm *object.VirtualMachine) (string, error) {
+	var o mo.VirtualMachine
+	err := vm.Properties(ctx, vm.Reference(), []string{"config.uuid"}, &o)
+	if err != nil {
+		return "", err
+	}
+
+	if o.Config != nil {
+		return o.Config.Uuid, nil
+	}
+
+	return "", fmt.Errorf("failed to get UUID for vm: %v", vm)
+}
+
+func (ops *vsphereOps) isVMValid(ctx context.Context, vm *object.VirtualMachine) (bool, error) {
+	var o mo.VirtualMachine
+	err := vm.Properties(ctx, vm.Reference(), []string{"runtime.connectionState"}, &o)
+	if err != nil {
+		return false, err
+	}
+
+	return o.Runtime.ConnectionState != "invalid", nil
+}
+
+func (ops *vsphereOps) getNameForVM(ctx context.Context, vm *object.VirtualMachine) (string, error) {
+	var o mo.VirtualMachine
+	err := vm.Properties(ctx, vm.Reference(), []string{"config.name"}, &o)
+	if err != nil {
+		return "", err
+	}
+
+	return o.Config.Name, nil
+}
+
+// We treat a VM's zone as the vSphere cluster in which the vm resizes.
+// If there is no cluster, then it's the datacenter
+func (ops *vsphereOps) getZoneForVM(
+	ctx context.Context,
+	vm *object.VirtualMachine) (string, error) {
+
+	host, err := vm.HostSystem(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	var o mo.HostSystem
+	err = host.Properties(ctx, host.Reference(), []string{"parent"}, &o)
+	if err != nil {
+		return "", err
+	}
+
+	var zone string
+	switch o.Parent.Type {
+	case "ClusterComputeResource":
+		ccr := object.NewClusterComputeResource(ops.conn.GoVmomiClient.Client, *o.Parent)
+		zone, err = ccr.ObjectName(ctx)
+		if err != nil {
+			return "", err
+		}
+		return zone, nil
+	case "ComputeResource":
+		// zone is the datacenter ( o -> parent -> parent -> parent )
+		cr := object.NewComputeResource(ops.conn.GoVmomiClient.Client, *o.Parent)
+		var o mo.ComputeResource
+		err = cr.Properties(ctx, cr.Reference(), []string{"parent"}, &o)
+		if err != nil {
+			return "", err
+		}
+
+		if o.Parent.Type == "Folder" {
+			folder := object.NewFolder(ops.conn.GoVmomiClient.Client, *o.Parent)
+			var o mo.Folder
+			err = folder.Properties(ctx, folder.Reference(), []string{"parent"}, &o)
+			if err != nil {
+				return "", err
+			}
+
+			if o.Parent.Type == "Datacenter" {
+				dc := object.NewDatacenter(ops.conn.GoVmomiClient.Client, *o.Parent)
+				zone, err = dc.ObjectName(ctx)
+				if err != nil {
+					return "", err
+				}
+				return zone, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("failed to get zone for vm: %s", vm.Name())
+}
+
+// addStorageToVM Current this just adds disk controllers. In future, add support
+// for iso, vmdks here
+func (ops *vsphereOps) addStorageToVM(
+	devices object.VirtualDeviceList,
+	controller string) (object.VirtualDeviceList, error) {
+	if controller != "ide" {
+		if controller == "nvme" {
+			nvme, err := devices.CreateNVMEController()
```

Comment:
CreateInstance passes in nil as "devices" argument. We probably need a nil check here?